### DISCUSSION TITLE
CRON

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#15444] [Q&A] [answered]
I notice that there is a CRON function available in Berry, is this exposed for use in the base esp32 code for a  sensor driver? 

Thanks....

**Comment by barbudor** [comment, marked as answer]
There are a few time related variables available to a driver that can be tested in the every-second call.

**Comment by trlafleur** [comment]
any reference to them available??

**Reply by barbudor** [reply]
Sorry not in front of my computer but I think they are in TasmotaGlobal structure.
There should be one that provides the minutes elapsed from midnight

**Comment by trlafleur** [comment]
Thanks, I will look for them....

Exposing the CRON function to a driver, or sensor code would be very
useful, in my case, I want to send a command to a device at a fixed
time...   for now, I parse the NTP time... this would take the load from
the 1-second call-back saving code run time...

This could be implemented in one or more standard call-backs to the
driver...

thanks

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
As I see it, cron is not "exposed" outside of Berry, meaning having a regular integration with the Tasmota main loop, but there is no security guard hindering you from working out a way of using it. I'd probably just opt for creating the driver in Berry.....

**Reply by barbudor** [reply]
I don't think having cron exposed would simplifie much the life of a driver or reduce load
As something will have to do the calculation, it would be centrelized instead of being in each driver.
That would not remove the calls to FUNC_EVERY_SECOND or others that are anyhow always "broadcasted" to every driver.

**Reply by s-hadinger** [reply]
Cron is based on `ccronexpr` available here `https://github.com/arendst/Tasmota/tree/development/lib/libesp32/ccronexpr`

Nothing prevents you from calling directly the C API of this lib in your code.

**Comment by trlafleur** [comment]
Thank you, I did find that and it's working just fine....

Where are the system time structures defined in Tasmota??, I'm having
difficulty locating them....

thanks...


~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by s-hadinger** [reply]
You can take a look at the Berry mapping as an example 

https://github.com/arendst/Tasmota/blob/development/lib/libesp32/berry_tasmota/src/be_cron_class.cpp
